### PR TITLE
Pitching in survey

### PR DESCRIPTION
**Pull request by ya7on**
# Please make sure all below checkboxes in the Mandatory section have been checked before submitting your PR (also don't forget to pay attention to the Depending on Circumstances section)

## Mandatory

- [x] The commit message is an imperative and starts with a capital letter (for example, `Add something`, `Remove something`, `Fix something`, etc.).
- [x] I made sure I hadn't put a period at the end of the the commit message(s).

## Depending on Circumstances

Some of the items in the section become mandatory if you are a first-time contributor and/or your changes are relevant to the items.

- [x] Adding a new parameter I didn't forget to reflect it in `README.md`.
- [ ] Modifying a list of constants or imports I didn't brake an alphabetical order.
- [ ] (for first-time contributors) One of the commits in the pull request adds me to the **Acknowledgements** section in `AUTHORS.md`.

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -124,6 +125,31 @@ exp.createBirthdayChannel = async (robot, username) => {
   robot.messageRoom(channelName, message)
 }
 
+/**
+ * Create a begging poll for all users except birthday user
+ *
+ * @param {Robot} robot - hubot instance
+ * @param {object} bdayUser - birthday boy/girl
+ */
+exp.createBeggingPoll = (robot, bdayUser) => {
```
Comment (eugulixes) [inline src/utils.js]: `createPitchingInSurvey`

File: src/utils.js [hunk 2]
```diff
@@ -124,6 +125,31 @@ exp.createBirthdayChannel = async (robot, username) => {
   robot.messageRoom(channelName, message)
 }
 
+/**
+ * Create a begging poll for all users except birthday user
```
Comment (eugulixes) [inline src/utils.js]: * ...a pitching in survey...
* except the birthday boy/girl
* put a dot (`.`) at the end of the sentence

File: src/birthder.js [hunk 3]
```diff
@@ -41,7 +41,9 @@
       delete: new RegExp(/(birthday delete)\s+/.source + regExpUsername.source + /\b/.source, 'i'),
       check: new RegExp(/(birthdays on)\s+/.source + regExpShortDate.source, 'i'),
       list: new RegExp(/(birthdays|fwd) list$/, 'i'),
-      fwd_set: new RegExp(/(fwd set)\s+/.source + regExpUsername.source + /\s+/.source + regExpDate.source, 'i')
+      fwd_set: new RegExp(/(fwd set)\s+/.source + regExpUsername.source + /\s+/.source + regExpDate.source, 'i'),
+      beggingAgree: new RegExp(/Yes. I can send money for gift for\s+/.source + regExpUsername.source, 'i'),
```
Comment (eugulixes) [inline src/birthder.js]: * `agreeToPitchIn`
* Yes, I'll pitch in on a present for...

File: src/birthder.js [hunk 4]
```diff
@@ -41,7 +41,9 @@
...
+      beggingReject: new RegExp(/No. I can not send money for\s+/.source + regExpUsername.source, 'i')
```
Comment (eugulixes) [inline src/birthder.js]: * `disagreeToPitchIn`
* No, I won't pitch in on a present for...

File: src/utils.js [hunk 5]
```diff
@@ -454,3 +485,29 @@ exp.detectBirthdaylessUsers = async robot => {
...
+      return exp.isEqualMonthDay(
+        moment(user.dateOfBirth, exp.DATE_FORMAT),
+        targetDay
+      )
+    })
+    .forEach(user => {
+      const all = users.length - 1
+      const agreed = user.birthdayBeggingList.length
+
+      const message = `Hi guys :wave: I was informed about the readiness of ${agreed} people out of ${all}!`
```
Comment (eugulixes) [inline src/utils.js]:
```
Good news, @all
${agreed} out of ${all} is/are going to pitch in.
```

File: src/utils.js [hunk 6]
```diff
@@ -454,3 +485,29 @@ exp.detectBirthdaylessUsers = async robot => {
     }
   }
 }
+
+/**
+ * Send count of users which are agreed for begging money
```
Comment (eugulixes) [inline src/utils.js]: Inform everybody in the birthday channel about the number of people who agreed to pitch in.

File: src/utils.js [hunk 7]
```diff
@@ -327,6 +353,7 @@ exp.removeExpiredBirthdayChannels = async (robot) => {
           channelName = userInstance.birthdayChannel.roomName
           await robot.adapter.api.post('groups.delete', { roomName: channelName })
           delete userInstance.birthdayChannel
+          delete userInstance.birthdayBeggingList
```
Comment (eugulixes) [inline src/utils.js]: `birthdayPitchingInList`

File: src/utils.js [hunk 8]
```diff
@@ -124,6 +125,31 @@ exp.createBirthdayChannel = async (robot, username) => {
...
+  return routines.getAllUsers(robot)
+    .then(users => {
+      users
+        .filter(user => user.id !== bdayUser.id)
+        .forEach(user => {
+          const text = `Hey. Soon the birthday of the ${bdayUser.name}. We collect money for his gift. Will you be able to participate in the collection of money? This is completely anonymous, we just need to know the amount.`
+          const message = routines.buildMessageWithButtons(
+            text,
+            [
+              ['Yes. I can', `Yes. I can send money for gift for ${bdayUser.name}`],
```
Comment (eugulixes) [inline src/utils.js]: `Yes, I'll pitch in on a present for ${bdayUser.name}`

File: src/utils.js [hunk 9]
```diff
@@ -124,6 +125,31 @@ exp.createBirthdayChannel = async (robot, username) => {
...
+              ['Sorry, but no', `No. I can not send money for ${bdayUser.name}`]
```
Comment (eugulixes) [inline src/utils.js]: `No, I won't pitch in on a present for ${bdayUser.name}`

File: src/utils.js [hunk 10]
```diff
@@ -454,3 +485,29 @@ exp.detectBirthdaylessUsers = async robot => {
     }
   }
 }
+
+/**
+ * Send count of users which are agreed for begging money
+ *
+ * @param {Robot} robot - hubot instance
```
Comment (eugulixes) [inline src/utils.js]: Put a dot (`.`) at the end of the line.

File: src/utils.js [hunk 11]
```diff
@@ -18,6 +18,7 @@ exp.COMPANY_NAME = process.env.COMPANY_NAME || 'WIS Software'
 exp.CREATE_BIRTHDAY_CHANNELS = process.env.CREATE_BIRTHDAY_CHANNELS === 'true' || false
 exp.HAPPY_REMINDER_SCHEDULER = process.env.HAPPY_REMINDER_SCHEDULER || '0 0 7 * * *'
 exp.NUMBER_OF_DAYS_IN_ADVANCE = parseInt(process.env.NUMBER_OF_DAYS_IN_ADVANCE, 10) || 7
+exp.CREATE_BEGGING_POLL = process.env.CREATE_BEGGING_POLL === 'true' || false
```
Comment (eugulixes) [inline src/utils.js]: `ENABLE_PITCHING_IN_SURVEY`

**Review by ABSLord — DISMISSED** [review]

File: README.md [hunk 12]
```diff
@@ -61,6 +61,7 @@ The script can be configured via the following environment variables (called par
 | `COMPANY_NAME` | Allows specifying the company name which is used in the welcome message. | WIS Software |
 | `HAPPY_REMINDER_SCHEDULER` | Allows specifying the frequency with which the script checks for nearest birthdays and writes birthday messages to users. The value of this parameter must follow the [Cron Format](https://github.com/node-schedule/node-schedule#cron-style-scheduling). | `0 0 7 * * *` |
 | `NUMBER_OF_DAYS_IN_ADVANCE`  | Sets (in days) how long before the event occurs the reminder will be triggered. | 7 |
+| `ENABLE_PITCHING_IN_SURVEY` | Specifies wether to create a pitching in survey for all users except the birthday boy/girl. | false |
```
Comment (eugulixes) [inline README.md]:
```
Specifies whether to create a pitching in survey. If the parameter is `true`, the bot asks every single users except the birthday boy/girl whether they will pitch in on a present for the birthday.
```

File: src/birthder.js [hunk 13]
```diff
@@ -235,6 +237,62 @@
...
+      const bdayUser = await routines.findUserByName(robot, targetUsername)
+
+      if (!bdayUser) {
+        msg.send(`I have never met ${targetUsername}.`)
+
+        return
+      }
+
+      if (!bdayUser.birthdayChannel) {
+        msg.send('This person does not plan to celebrate his birthday in the near future. Do not rush.')
```
Comment (eugulixes) [inline src/birthder.js]:
```
This person does not plan to celebrate their birthday in the nears future. Do not rush.
```

File: src/birthder.js [hunk 14]
```diff
@@ -235,6 +237,62 @@
...
```
Comment (eugulixes) [inline src/birthder.js]:
```
This person does not plan to celebrate their birthday in the nears future. Do not rush.
```

File: src/birthder.js [hunk 15]
```diff
@@ -235,6 +237,62 @@
...
+      }
+
+      bdayUser.birthdayPitchingInList = bdayUser.birthdayPitchingInList || []
+
+      if (bdayUser.birthdayPitchingInList.includes(msg.message.user.id)) {
+        bdayUser.birthdayPitchingInList = bdayUser.birthdayPitchingInList.filter(
+          userId => userId !== msg.message.user.id
+        )
+
+        msg.send(`You already tell me that you will. I removed you from the list.`)
```
Comment (eugulixes) [inline src/birthder.js]:
```
It's a shame. I will remove you from the pitching in list.
```

File: src/birthder.js [hunk 16]
```diff
@@ -235,6 +237,62 @@
...
+      } else {
+        msg.send('Very sorry. But i memorized your choice.')
```
Comment (eugulixes) [inline src/birthder.js]:
```
Ok, I understand.
```

File: src/utils.js [hunk 17]
```diff
@@ -124,6 +125,31 @@ exp.createBirthdayChannel = async (robot, username) => {
...
+ * @param {Robot} robot - hubot instance
+ * @param {object} bdayUser - birthday boy/girl
+ */
+exp.createPitchingInSurvey = (robot, bdayUser) => {
+  return routines.getAllUsers(robot)
+    .then(users => {
+      users
+        .filter(user => user.id !== bdayUser.id)
+        .forEach(user => {
+          const text = `Hey. Soon the birthday of the ${bdayUser.name}. We collect money for his gift. Will you be able to participate in the collection of money? This is completely anonymous, we just need to know the amount.`
```
Comment (eugulixes) [inline src/utils.js]:
```
Hello. ${bdayUser.name} is having a birthday soon. We pitching in on the present. Are you in?
This is a completely anonymous survey. We simply need to know the number of the people who are going to pitch in.
```

File: src/birthder.js [hunk 18]
```diff
@@ -235,6 +237,62 @@
...
+      if (!bdayUser.birthdayChannel) {
+        msg.send('This person does not plan to celebrate his birthday in the near future. Do not rush.')
+
+        return
+      }
+
+      bdayUser.birthdayPitchingInList = bdayUser.birthdayPitchingInList || []
+
+      if (bdayUser.birthdayPitchingInList.includes(msg.message.user.id)) {
+        msg.send(`You already tell me that you will. I left everything as it was.`)
```
Comment (eugulixes) [inline src/birthder.js]:
```
You have already told me that you would.
```

File: README.md [hunk 19]
```diff
@@ -61,6 +61,7 @@ The script can be configured via the following environment variables (called par
 | `COMPANY_NAME` | Allows specifying the company name which is used in the welcome message. | WIS Software |
 | `HAPPY_REMINDER_SCHEDULER` | Allows specifying the frequency with which the script checks for nearest birthdays and writes birthday messages to users. The value of this parameter must follow the [Cron Format](https://github.com/node-schedule/node-schedule#cron-style-scheduling). | `0 0 7 * * *` |
 | `NUMBER_OF_DAYS_IN_ADVANCE`  | Sets (in days) how long before the event occurs the reminder will be triggered. | 7 |
+| `ENABLE_PITCHING_IN_SURVEY` | Specifies whether to create a pitching in survey. If the parameter is `true`, the bot asks every single users except the birthday boy/girl whether they will pitch in on a present for the birthday. | false |
```
Comment (eugulixes) [inline README.md]: `every single user`

File: src/birthder.js [hunk 20]
```diff
@@ -235,6 +237,62 @@
...
+      const bdayUser = await routines.findUserByName(robot, targetUsername)
+
+      if (!bdayUser) {
+        msg.send(`I have never met ${targetUsername}.`)
+
+        return
+      }
+
+      if (!bdayUser.birthdayChannel) {
+        msg.send('This person does not plan to celebrate their birthday in the nears future. Do not rush.')
```
Comment (eugulixes) [inline src/birthder.js]: `nearest future`

File: src/birthder.js [hunk 21]
```diff
@@ -235,6 +237,62 @@
...
```
Comment (eugulixes) [inline src/birthder.js]: `nearst future`

File: src/utils.js [hunk 22]
```diff
@@ -124,6 +125,31 @@ exp.createBirthdayChannel = async (robot, username) => {
...
+ * @param {Robot} robot - hubot instance
+ * @param {object} bdayUser - birthday boy/girl
+ */
+exp.createPitchingInSurvey = (robot, bdayUser) => {
+  return routines.getAllUsers(robot)
+    .then(users => {
+      users
+        .filter(user => user.id !== bdayUser.id)
+        .forEach(user => {
+          const text = `Hello. ${bdayUser.name} is having a birthday soon. We pitching in on the present. Are you in? This is a completely anonymous survey. We simply need to know the number of the people who are going to pitch in.`
```
Comment (eugulixes) [inline src/utils.js]: `We are pitching in on the present`

**Review by eugulixes — APPROVED** [review]

**Pull request merged**